client/src/js/components/bhCheckboxTree/bhCheckboxTree.js

```diff
@@ -0,0 +1,158 @@
+angular.module('bhima.components')
+  .component('bhCheckboxTree', {
+    templateUrl : 'js/components/bhCheckboxTree/bhCheckboxTree.html',
+    controller  : bhCheckboxTreeController,
+    bindings    : {
+      data : '<',
+      checkedIds : '<?',
+      onChange : '&',
+      idKey : '@?',
+      labelKey : '@?',
+      parentKey : '@?',
+      isFlatTree : '@?',
+    },
+  });
+
+bhCheckboxTreeController.$inject = ['TreeService'];
+
+function bhCheckboxTreeController(Tree) {
+  const $ctrl = this;
+
+  const DEFAULT_ID_KEY = 'id';
+  const DEFAULT_LABEL_KEY = 'label';
+  const DEFAULT_PARENT_KEY = 'parent';
+
+  $ctrl.$onInit = () => {
+    $ctrl.idKey = $ctrl.idKey || DEFAULT_ID_KEY;
+    $ctrl.labelKey = $ctrl.labelKey || DEFAULT_LABEL_KEY;
+    $ctrl.parentKey = $ctrl.parentKey || DEFAULT_PARENT_KEY;
+
+    if ($ctrl.data) {
+      buildTree($ctrl.data);
+    }
+  };
+
+  $ctrl.$onChanges = (changes) => {
+    if (changes.data && changes.data.currentValue) {
+      buildTree(changes.data.currentValue);
+    }
+  };
+
+  function buildTree(array = []) {
+    const data = [...array];
+
+    // make easier template labels and default values to checked is false
+    data.forEach(node => {
+      node._label = node[$ctrl.labelKey];
+      node._checked = false;
+
+      // work on flat arrays by faking a tree
+      if ($ctrl.isFlatTree) { node[$ctrl.parentKey] = 0; }
+    });
+
+    // create the tree
+    $ctrl.tree = new Tree(data, { parentKey : $ctrl.parentKey, rootId : 0, idKey : $ctrl.idKey });
+    $ctrl.root = $ctrl.tree.getRootNode();
+
+    // compute node depths
+    $ctrl.tree.walk(Tree.common.computeNodeDepth);
+
+    // ensure that checked ids are an array
+    if (!Array.isArray($ctrl.checkedIds) || data.length === 0) {
+      return;
+    }
+
+    // ensure that the mask is a cloned array
+    const mask = [...$ctrl.checkedIds];
```

Comment:
Is this a nowadays JSON.parse(JSON.stringify(array)) ? for having a new array which doesn't have the same reference to the old one

Reply:
Correct.  Although it is more like something like:
```js
var a = [1, 2, 3];
var b = a.slice();
console.log(a !== b);
```